consensus/src/liveness/proposal_status_tracker.rs

```diff
@@ -0,0 +1,145 @@
+use super::round_state::NewRoundReason;
+use aptos_collections::BoundedVecDeque;
+use aptos_consensus_types::{
+    common::Author, payload_pull_params::OptQSPayloadPullParams, round_timeout::RoundTimeoutReason,
+};
+use aptos_infallible::Mutex;
+use std::{collections::HashSet, sync::Arc};
+
+pub trait TPastProposalStatusTracker: Send + Sync {
+    fn push(&self, status: NewRoundReason);
+}
+
+pub trait TOptQSPullParamsProvider: Send + Sync {
+    fn get_params(&self) -> Option<OptQSPayloadPullParams>;
+}
+
+/// A exponential window based algorithm to decide whether to go optimistic or not.
+///
+/// Initialize the window at 2.
+/// - For each proposal failure, double the window up to a MAX size
+/// - If there are no failures within the window, the propose optimistic batch
+/// - If there are no failures up to MAX proposals, reset the window to 2.
+pub struct ExponentialWindowFailureTracker {
+    window: usize,
+    max_window: usize,
+    past_round_statuses: BoundedVecDeque<NewRoundReason>,
+    last_consecutive_success_count: usize,
+    ordered_authors: Vec<Author>,
+}
+
+impl ExponentialWindowFailureTracker {
+    pub(crate) fn new(max_window: usize, ordered_authors: Vec<Author>) -> Self {
+        Self {
+            window: 2,
+            max_window,
+            past_round_statuses: BoundedVecDeque::new(max_window),
+            last_consecutive_success_count: 0,
+            ordered_authors,
+        }
+    }
+
+    pub(crate) fn push(&mut self, status: NewRoundReason) {
+        self.past_round_statuses.push_back(status);
+        self.compute_failure_window();
+    }
+
+    fn last_consecutive_statuses_matching<F>(&self, matcher: F) -> usize
+    where
+        F: Fn(&NewRoundReason) -> bool,
+    {
+        let mut last_consecutive_matching_count = 0;
+        for reason in self.past_round_statuses.iter().rev() {
+            if matcher(reason) {
+                last_consecutive_matching_count += 1;
+            } else {
+                break;
+            }
+        }
+        last_consecutive_matching_count
+    }
+
+    fn compute_failure_window(&mut self) {
+        self.last_consecutive_success_count =
+            self.last_consecutive_statuses_matching(|reason| match reason {
+                NewRoundReason::QCReady => true,
```

Comment:
I think you can write like this if we don't care about listing all other reasons
```
NewRoundReason::Timeout(RoundTimeoutReason::PayloadUnavailable) => false,
_ => true,
```